src/vmm/src/arch/aarch64/mod.rs

```diff
@@ -58,9 +58,24 @@ pub const MMIO_MEM_SIZE: u64 = layout::DRAM_MEM_START - layout::MAPPED_IO_START;
 
 /// Returns a Vec of the valid memory addresses for aarch64.
 /// See [`layout`](layout) module for a drawing of the specific memory model for this platform.
-pub fn arch_memory_regions(size: usize) -> Vec<(GuestAddress, usize)> {
-    let dram_size = min(size, layout::DRAM_MEM_MAX_SIZE);
-    vec![(GuestAddress(layout::DRAM_MEM_START), dram_size)]
+///
+/// The `offset` parameter specified the offset from [`layout::DRAM_MEM_START`].
+pub fn arch_memory_regions(offset: usize, size: usize) -> Vec<(GuestAddress, usize)> {
+    assert!(size > 0, "Attempt to allocate guest memory of length 0");
+    assert!(
+        offset.checked_add(size).is_some(),
+        "Attempt to allocate guest memory such that the address space would wrap around"
+    );
```

Comment:
I would also add a check that `offset + size <= layout::DRAM_MEX_MAX_SIZE`

Reply:
I deliberately didn't add this because the existing behavior of this function before this commit was "if size > DRAM_MAX_MEM_SIZE then truncate to DRAM_MAX_MEM_SIZE", so that's why I only added the assert for "offset < DRAM_MAX_MEM_SIZE" (which is needed to ensure that after truncation we still have a non-zero memory region)

Reply:
My point is that maybe the behaviour of this function should not be that. Or, at the very least it would need to warn users that truncation is happening, so we can help them debugging issues.